roledb.go

```diff
@@ -289,23 +294,43 @@ func ReadRawGroups(importedGroups []RawAuthGroup) ([]AuthGroup, error) {
 	return groups, nil
 }
 
-// GroupIDsToName converts group IDs to names.
+// GroupIDsToNames converts group IDs to names.
 // The 'cache' parameter is used to speed up subsequent calls to this function,
-// because this function tends to get used in loops.
+// because this function tends to get used in loops. This function does not remove items from the cache - cache
+// management is left to the consumer.
 func GroupIDsToNames(groups []GroupIDU32, db RoleGroupDB, cache map[GroupIDU32]string) ([]string, error) {
-	names := make([]string, len(groups))
-	for i, gid := range groups {
+	names := make([]string, 0, len(groups))
+	var groupsDB []*AuthGroup
+	for _, gid := range groups {
 		if cache[gid] != "" {
-			names[i] = cache[gid]
+			names = append(names, cache[gid])
 		} else {
-			if group, err := db.GetByID(gid); err != nil {
+			// why is this not retrieved in one go? the below would result in 'n' nr of individual queries to db
+			// the below should only happen at most once per request
+			// if no new groups are added before the next request, cache should be up to date
+			// CAVEAT - For new groups, the groups will be loaded ONCE per function call if a new group detected,
+			// but for unknown groups, this will ALWAYS happen, since it will never be found in the GroupsDB.
+			var err error
+			if groupsDB == nil {
+				groupsDB, err = db.GetGroups()
+			}
+			if err != nil {
 				return nil, err
 			} else {
-				names[i] = group.Name
-				cache[gid] = group.Name
+				// re-populate groups cache
+				for _, giddb := range groupsDB {
+					if cache[giddb.ID] == "" {
+						cache[giddb.ID] = giddb.Name
+					}
+				}
+				// cycle through new groups, find in cache (based on new group id)
+				if cache[gid] != "" {
+					names = append(names, cache[gid])
+				}
 			}
```

Comment:
I really don't like all of this logic living in an `else` statement... Most of the lines in this function now have nothing to do with the function... Please create the function I asked for, even if you have to pass in the `cache` as a parameter to the function.

Also, we still need to return an error, you can take the same approach as you did with `PermitResolveToList` and just keep the error around to return at the very end. You will most likely have to write the error message your self.

As to your comment, I would remove the first 3 lines, as it only makes sens if the developer was also looking at the old code. But the last two lines I feel is important in case we ever have performance issues and someone needs to investigate.

Reply:
I think between this comment and the previous one on the older code we've missed each other. Did you mean "the calling function should _populate_ the cache" instead? Since the calling function _does_ initialise the cache...
And then - would you prefer we keep the old code and provide a _new_ function as an enhancement (and replace some of the new code in the new function with another function so as to simplify the logic there)?
I'll do what i think you want and perhaps discuss offline if still required.